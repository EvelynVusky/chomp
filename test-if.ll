; ModuleID = 'Chomp'
source_filename = "Chomp"

@fmt = private unnamed_addr constant [3 x i8] c"%d\00", align 1
@fmt.1 = private unnamed_addr constant [3 x i8] c"%s\00", align 1
@fmt.2 = private unnamed_addr constant [3 x i8] c"%c\00", align 1
@fmt.3 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@fmt.4 = private unnamed_addr constant [30 x i8] c"cannot call car on empty list\00", align 1
@string = private unnamed_addr constant [4 x i8] c"yes\00", align 1
@string.5 = private unnamed_addr constant [5 x i8] c"true\00", align 1
@string.6 = private unnamed_addr constant [6 x i8] c"false\00", align 1
@string.7 = private unnamed_addr constant [3 x i8] c"no\00", align 1

declare i32 @printf(i8*, ...)

declare i32 @print_bit(i32)

declare i32 @print_nibble(i32)

declare i32 @print_byte(i32)

declare i32 @print_word(i32)

declare i32 @exit(i32, ...)

define i32 @main() {
entry:
  %i = alloca i32, align 4
  store i32 0, i32* %i, align 4
  %b = alloca i1, align 1
  store i1 true, i1* %b, align 1
  %i1 = load i32, i32* %i, align 4
  %tmp = icmp eq i32 %i1, 0
  br i1 %tmp, label %then, label %else10

merge:                                            ; preds = %else10, %merge4
  ret i32 0

then:                                             ; preds = %entry
  %printf = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @fmt.1, i32 0, i32 0), i8* getelementptr inbounds ([4 x i8], [4 x i8]* @string, i32 0, i32 0))
  %printf2 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @fmt.3, i32 0, i32 0))
  %b3 = load i1, i1* %b, align 1
  br i1 %b3, label %then5, label %else

merge4:                                           ; preds = %else, %then5
  br label %merge

then5:                                            ; preds = %then
  %printf6 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @fmt.1, i32 0, i32 0), i8* getelementptr inbounds ([5 x i8], [5 x i8]* @string.5, i32 0, i32 0))
  %printf7 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @fmt.3, i32 0, i32 0))
  br label %merge4

else:                                             ; preds = %then
  %printf8 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @fmt.1, i32 0, i32 0), i8* getelementptr inbounds ([6 x i8], [6 x i8]* @string.6, i32 0, i32 0))
  %printf9 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @fmt.3, i32 0, i32 0))
  br label %merge4

else10:                                           ; preds = %entry
  %printf11 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @fmt.1, i32 0, i32 0), i8* getelementptr inbounds ([3 x i8], [3 x i8]* @string.7, i32 0, i32 0))
  %printf12 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([2 x i8], [2 x i8]* @fmt.3, i32 0, i32 0))
  br label %merge
}
