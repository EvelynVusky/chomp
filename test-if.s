	.section	__TEXT,__text,regular,pure_instructions
	.build_version macos, 14, 0
	.globl	_main                           ## -- Begin function main
	.p2align	4, 0x90
_main:                                  ## @main
	.cfi_startproc
## %bb.0:                               ## %entry
	pushq	%rax
	.cfi_def_cfa_offset 16
	movl	$0, 4(%rsp)
	movb	$1, 3(%rsp)
	movb	$1, %al
	leaq	L_fmt.1(%rip), %rdi
	testb	%al, %al
	jne	LBB0_1
## %bb.5:                               ## %else10
	leaq	L_string.8(%rip), %rsi
	jmp	LBB0_3
LBB0_1:                                 ## %then
	leaq	L_string(%rip), %rsi
	xorl	%eax, %eax
	callq	_printf
	leaq	L_fmt.3(%rip), %rdi
	xorl	%eax, %eax
	callq	_printf
	cmpb	$1, 3(%rsp)
	jne	LBB0_4
## %bb.2:                               ## %then5
	leaq	L_fmt.1(%rip), %rdi
	leaq	L_string.6(%rip), %rsi
	jmp	LBB0_3
LBB0_4:                                 ## %else
	leaq	L_fmt.1(%rip), %rdi
	leaq	L_string.7(%rip), %rsi
LBB0_3:                                 ## %merge
	xorl	%eax, %eax
	callq	_printf
	leaq	L_fmt.3(%rip), %rdi
	xorl	%eax, %eax
	callq	_printf
	xorl	%eax, %eax
	popq	%rcx
	retq
	.cfi_endproc
                                        ## -- End function
	.section	__TEXT,__cstring,cstring_literals
L_fmt:                                  ## @fmt
	.asciz	"%d"

L_fmt.1:                                ## @fmt.1
	.asciz	"%s"

L_fmt.2:                                ## @fmt.2
	.asciz	"%c"

L_fmt.3:                                ## @fmt.3
	.asciz	"\n"

L_fmt.4:                                ## @fmt.4
	.asciz	"Error: cannot call car on empty list"

L_fmt.5:                                ## @fmt.5
	.asciz	"Error: cannot call cdr on empty list"

L_string:                               ## @string
	.asciz	"yes"

L_string.6:                             ## @string.6
	.asciz	"true"

L_string.7:                             ## @string.7
	.asciz	"false"

L_string.8:                             ## @string.8
	.asciz	"no"

.subsections_via_symbols
