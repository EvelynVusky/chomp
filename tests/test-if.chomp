
void foo (int a) {
    if (a > 4) {
        println("greater");
        if (a == 5) {
            println("greater yes");
        } else {
            println("greater no");
        }
    } else {
        println("less");
        if (a == 3) {
            println("less yes");
        } else {
            println("less no");
        }
    }

}

int main () {
    foo(5);
    foo(6);
    foo(3);
    foo(2);

    int a = 7;

    if (a > 4) {
        println("greater");
        if (a == 5) {
            println("greater yes");
        } else {
            println("greater no");
        }
    } else {
        println("less");
        if (a == 3) {
            println("less yes");
        } else {
            println("less no");
        }
    }
    return 0;
}