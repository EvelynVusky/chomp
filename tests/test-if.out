greater
greater yes
greater
greater no
less
less yes
less
less no
greater
greater no
